### ISSUE TITLE
[STORM-821] storm-jdbc create a connection provider interface to decouple from hikariCP being the only connection pool implementation that can be used.

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#4662]
<p>The current implementation of storm-jdbc is couple with HikariCP configuration. We propose to remove this coupling by introducing a connectionProvider interface with a default HikariCP implementation. This will allow users to do their own connection pool management or chose a different connection pooling library. </p>

---
<details><summary><i>Originally reported by <a title="Parth Brahmbhatt" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt">parth.brahmbhatt</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-821" target="_blank">storm-jdbc create a connection provider interface to decouple from hikariCP being the only connection pool implementation that can be used.</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Parth Brahmbhatt" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=parth.brahmbhatt">parth.brahmbhatt</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2015-06-02T19:50:41+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user Parth-Brahmbhatt opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-821" title="storm-jdbc create a connection provider interface to decouple from hikariCP being the only connection pool implementation that can be used." class="issue-link" data-issue-key="STORM-821"><del>STORM-821</del></a>: Adding connection provider interface to decouple jdbc conn…</p>

<p>    …ector from a single connection pooling implementation.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/Parth-Brahmbhatt/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Parth-Brahmbhatt/incubator-storm</a> <a href="https://issues.apache.org/jira/browse/STORM-821" title="storm-jdbc create a connection provider interface to decouple from hikariCP being the only connection pool implementation that can be used." class="issue-link" data-issue-key="STORM-821"><del>STORM-821</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/556.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #556</p>

<hr />
<p>commit 94e4c2ad471bb5301820b9a36e36d0ba43a12d0f<br/>
Author: Parth Brahmbhatt <brahmbhatt.parth@gmail.com><br/>
Date:   2015-05-15T23:32:44Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-821" title="storm-jdbc create a connection provider interface to decouple from hikariCP being the only connection pool implementation that can be used." class="issue-link" data-issue-key="STORM-821"><del>STORM-821</del></a>: Adding connection provider interface to decouple jdbc connector from a single connection pooling implementation.</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-103227689" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-103227689</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105621487" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105621487</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105689247" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105689247</a></p>

<p>    Class name has a typo: ConnectionPrvoider -> ConnectionProvider.<br/>
    It's not just documentation typo, but also real classname typo.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#discussion_r31087238" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#discussion_r31087238</a></p>

<p>    &#8212; Diff: external/storm-jdbc/README.md &#8212;<br/>
    @@ -17,11 +44,12 @@ The `getColumns()` method defines how a storm tuple maps to a list of columns re<br/>
     *<b>The order of the returned list is important. The place holders in the supplied queries are resolved in the same order as returned list.</b>*<br/>
     For example if the user supplied insert query is `insert into user(user_id, user_name, create_date) values (?,?, now())` the 1st item <br/>
     of the returned list of `getColumns` method will map to the 1st place holder and the 2nd to the 2nd and so on. We do not parse<br/>
    -the supplied queries to try and resolve place holder by column names. <br/>
    +the supplied queries to try and resolve place holder by column names. Not making any assumptions about the query syntax allows this connector<br/>
    +to be used by some non-standard sql frameworks like Pheonix which only supports upsert into.</p>

<ol>
	<li>
	<ol>
		<li>
		<ol>
			<li>JdbcInsertBolt<br/>
    -To use the `JdbcInsertBolt`, you construct an instance of it and specify a configuration key in your storm config that holds the <br/>
    -hikari configuration map and a `JdbcMapper` implementation that coverts storm tuple to DB row. In addition, you must either supply <br/>
    +To use the `JdbcInsertBolt`, you construct an instance of it by specifying a `ConnectionProvider` implementation<br/>
    +and a `JdbcMapper` implementation that coverts storm tuple to DB row. In addition, you must either supply</li>
		</ol>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ol>
	</li>
</ol>


<p>    coverts -> converts</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#discussion_r31087567" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#discussion_r31087567</a></p>

<p>    &#8212; Diff: external/storm-jdbc/README.md &#8212;<br/>
    @@ -29,13 +57,21 @@ The default is set to value of topology.message.timeout.secs and a value of -1 w<br/>
     You should set the query timeout value to be <= topology.message.timeout.secs.</p>

<p>      ```java<br/>
    -Config config = new Config();<br/>
    -config.put("jdbc.conf", hikariConfigMap);<br/>
    -JdbcInsertBolt userPersistanceBolt = new JdbcInsertBolt("jdbc.conf",simpleJdbcMapper)<br/>
    +Map hikariConfigMap = Maps.newHashMap();<br/>
    &#8212; End diff &#8211;</p>

<p>    minor thing: <br/>
    How about changing this line to <br/>
    ```<br/>
    Map<String, String> hikariConfigMap = new HashMap<String, String>();<br/>
    ```<br/>
    to remove dependency about Guava from example?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#discussion_r31087884" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#discussion_r31087884</a></p>

<p>    &#8212; Diff: external/storm-jdbc/src/main/java/org/apache/storm/jdbc/mapper/SimpleJdbcMapper.java &#8212;<br/>
    @@ -33,9 +34,10 @@</p>

<p> private List<Column> schemaColumns;</p>

<ul class="alternate" type="square">
	<li>public SimpleJdbcMapper(String tableName, Map hikariConfigurationMap) {<br/>
    +    public SimpleJdbcMapper(String tableName, Map<String, Object> connectionProvideConfig, ConnectionProvider connectionProvider) {
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Seems like we don't need to connectionProvideConfig here for now.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#discussion_r31087950" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#discussion_r31087950</a></p>

<p>    &#8212; Diff: external/storm-jdbc/README.md &#8212;<br/>
    @@ -60,6 +96,7 @@ hikariConfigMap.put("dataSourceClassName","com.mysql.jdbc.jdbc2.optional.MysqlDa<br/>
     hikariConfigMap.put("dataSource.url", "jdbc:mysql://localhost/test");<br/>
     hikariConfigMap.put("dataSource.user","root");<br/>
     hikariConfigMap.put("dataSource.password","password");<br/>
    +ConnectionProvider connectionProvider = new HikariCPConnectionProvider(map);<br/>
     String tableName = "user_details";<br/>
     JdbcMapper simpleJdbcMapper = new SimpleJdbcMapper(tableName, map);<br/>
    &#8212; End diff &#8211;</p>

<p>    After modifying <a href="https://github.com/apache/storm/pull/556/files#r31087884" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556/files#r31087884</a>, it should be changed to <br/>
    ```<br/>
    JdbcMapper simpleJdbcMapper = new SimpleJdbcMapper(tableName, connectionProvider);<br/>
    ```</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105691986" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105691986</a></p>

<p>    @HeartSaVioR  all comments addressed, thanks for finding the README issues and keeping the documentation honest. I did not remove the Guava dependency as there are multiple references to Lists and I think Guava is so widely used its unlikely to confuse anyone reading the README.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user Parth-Brahmbhatt commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105692443" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105692443</a></p>

<p>    @harshach @ptgoetz @HeartSaVioR I forgot to mention that I also made a minor change in column.java so we can handle database columns with null values. It was a small change so filing another jira felt like an overkill. If anyone has objection to that let me know and I will open a different jira.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105693031" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105693031</a></p>

<p>    @Parth-Brahmbhatt I agree that Guava is widely-used library. I'm OK to keep dependency from example. <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#discussion_r31089108" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#discussion_r31089108</a></p>

<p>    &#8212; Diff: external/storm-jdbc/README.md &#8212;<br/>
    @@ -60,6 +96,7 @@ hikariConfigMap.put("dataSourceClassName","com.mysql.jdbc.jdbc2.optional.MysqlDa<br/>
     hikariConfigMap.put("dataSource.url", "jdbc:mysql://localhost/test");<br/>
     hikariConfigMap.put("dataSource.user","root");<br/>
     hikariConfigMap.put("dataSource.password","password");<br/>
    +ConnectionProvider connectionProvider = new HikariCPConnectionProvider(map);<br/>
     String tableName = "user_details";<br/>
     JdbcMapper simpleJdbcMapper = new SimpleJdbcMapper(tableName, map);<br/>
    &#8212; End diff &#8211;</p>

<p>    @Parth-Brahmbhatt This line is not changed yet, so you may want to check it again.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-105697874" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-105697874</a></p>

<p>    LGTM. :+1: </p>

<p>    Btw, we don't release storm-jdbc module to any Storm releases yet, so IMO it could be not mandatory to have a new issue regarding to Column when we promise #556 will be merged to 0.10.x before releasing.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user asfgit closed the pull request at:</p>

<p>    <a href="https://github.com/apache/storm/pull/556" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/556#issuecomment-106914408" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/556#issuecomment-106914408</a></p>

<p>    Thanks @Parth-Brahmbhatt. I merged this into master and 0.10.x.</p>
